src/win/pipe.c

```diff
@@ -1782,6 +1782,88 @@ int uv__pipe_write(uv_loop_t* loop,
 }
 
 
+int uv__pipe_try_write(uv_pipe_t* handle,
+                       const uv_buf_t bufs[],
+                       unsigned int nbufs) {
+  OVERLAPPED overlapped;
+  HANDLE event;
+  const uv_buf_t* buf;
+  int bytes_written;
+  unsigned int idx;
+  DWORD timeout;
+  DWORD err;
+
+  if (handle->flags & UV_HANDLE_NON_OVERLAPPED_PIPE) {
+    return UV_EAGAIN;
+  }
+
+  if (handle->stream.conn.write_reqs_pending > 0) {
+    return UV_EAGAIN;
+  }
+
+  timeout = 0;
+  if (handle->flags & UV_HANDLE_BLOCKING_WRITES) {
+    timeout = INFINITE;
+  }
+
+  memset(&overlapped, 0, sizeof(overlapped));
+
+  event = CreateEvent(NULL, FALSE, FALSE, NULL);
+  if (event == NULL) {
+    uv_fatal_error(GetLastError(), "CreateEvent");
+  }
+
+  overlapped.hEvent = (HANDLE)((uintptr_t)event | 1);
+
+  bytes_written = 0;
+  for (err = 0, idx = 0; idx < nbufs; err = 0, idx += 1) {
+    buf = &bufs[idx];
+
+    if (WriteFile(handle->handle, buf->base, buf->len, NULL, &overlapped)) {
+      bytes_written += buf->len;
+      continue;
+    }
+
+    err = GetLastError();
+    if (err != ERROR_IO_PENDING) {
+      break;
+    }
+
+    err = WaitForSingleObject(event, timeout);
+    if (err == WAIT_OBJECT_0) {
+      bytes_written += buf->len;
+      continue;
+    }
+
+    if (err == WAIT_TIMEOUT && CancelIo(handle->handle)) {
+      if (GetOverlappedResult(handle->handle, &overlapped, &err, TRUE)) {
```

Comment:
this should probably be called whether or not `CancelIOEx` succeeds, to ensure we don't return from this function (freeing the `overlapped` memory) while the kernel is still using it (the TRUE here guarantees that, along with the use of the unique CreateEvent kernel handle)

Though this could potentially use handle->handle instead as the newly allocated kernel event HANDLE, and then loop here while ERROR_IO_PENDING is returned? The documentation says not to do that, but that feels like it must be wrong.

Reply:
> this should probably be called whether or not `CancelIOEx` succeeds, to ensure we don't return from this function (freeing the `overlapped` memory) while the kernel is still using it (the TRUE here guarantees that, along with the use of the unique CreateEvent kernel handle)
> 
No windows expert here, but by looking at the [documentation](https://learn.microsoft.com/en-us/windows/win32/fileio/cancelioex-func#return-value) I'm not sure that's an issue. If I read this correctly, if `CancelIOEx` fails the cancellation request wasn't successful: can the kernel still be using it at this point?. Also, it only mentions using `GetOverlappedResult()` in case of success.

Reply:
It is undocumented why it might fail, but I would guess the most likely code is ERROR_NOT_SUPPORTED from it if the underlying driver does not implement this functionality. That does not apply to most NamedPipe objects (which would be using the native named pipe drive), but could possibly apply if the user created their own `fd` that was namedpipe-like. If it errored in this way, the OVERLAPPED operation would indeed still remain pending after the error (since it was successfully started that way via the WriteFile call). Indeed, the documentation clarifies that the OVERLAPPED operation is still also pending after the call succeeds, but it is more likely to stop soon in that case.